### ISSUE TITLE
Socket.io keeps disconnecting and reconnecting

### ISSUE THREAD

**Issue by wenzhiwang** [react-community/create-react-native-app#133]
## Description

I'm using socket.io client in my react-native app created by CRNA. On my iPhone, Expo app successfully connected to my local server using socket.io, but it drops connection immediately. It then keeps reconnecting and disconnecting. 

## Expected Behavior

It should stay connected. It worked with Facebook's code example using the build in Websocket
Client:
https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/js/WebSocketExample.js
Server:
https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/js/websocket_test_server.js

## Observed Behavior

socket.io connection keeps disconnecting. However, it works fine when I run the same code on my Android simulator. 
## Environment

Please run these commands in the project folder and fill in their results:

* `npm ls react-native-scripts`: react-native-scripts@0.0.25
* `npm ls react-native`: `-- react-native@0.42.3
* `npm ls expo`:`-- expo@15.1.2
* `node -v`: v7.7.2
* `npm -v`: 3.10.9
* `yarn --version`: v0.21.3
* `watchman version`:

1. Operating system: Windows
2. Phone/emulator/simulator & version: iPhone 7+

## Reproducible Demo
```javascript
window.navigator.userAgent = "react-native";
const io = require('socket.io-client/dist/socket.io');
const socket = io('http://192.168.0.105:3001', {
  jsonp: false,
  transports: ['websocket'] // you need to explicitly tell it to use websockets
});
  constructor(props) {
    super(props);
    socket.on('connect', () => {
      socket.emit('message', "test");
    })
  }
```

**Comment by dikaiosune** [comment]
Thanks for the report!

At first glance I would assume that this is a networking problem. If it is networking, I wouldn't be surprised that your local emulator is able to stay connected since it doesn't have to traverse your WiFi. If you run similar code in your phone's browser, do you experience similar issues?

**Comment by wenzhiwang** [comment]
Thanks for the response.
It could be, but it works fine for Facebook's example code with react-native's build in Websocket. 
Client:
https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/js/WebSocketExample.js
Server:
https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/js/websocket_test_server.js
The communication between the app running on Expo and my PC stay connected. I will try to run a web app that connects to the server tonight.

**Comment by wenzhiwang** [comment]
@dikaiosune 
I tried to run a web app on my iPhone, it connects to the express server on my PC successfully, and it doesn't drop the connection. Run the react native app with same configuration on Expo, it still drops connections. 
Server side
```javascript

const server = app.listen(process.env.PORT, '0.0.0.0', function(err) {
  if (err) {
    console.log(err);
    return;
  }
  console.log('server listening on port: %s', process.env.PORT);
});
'

**Comment by dikaiosune** [comment]
Ah, I see -- so regular websockets work fine but socket.io does not? It looks like there's an open issue in socket.io regarding a similar-sounding problem with using the client in React Native:

https://github.com/socketio/socket.io/issues/2565

Specifically:

https://github.com/socketio/socket.io/issues/2565#issuecomment-256835129

Can you see if the diagnosis in the initial comment is true for your project as well?

**Comment by dikaiosune** [comment]
Also, you may want to take a look at this example that @brentvatne made: https://github.com/expo/socket-io-example

**Comment by wenzhiwang** [comment]
@dikaiosune  Thanks a lot! ngrok is life saver. The only thing I need to do is to use ngrok.
I don't know why socket.io has connection problem with my local server through WiFi while Websocket works fine. Maybe because I have to use https on iOS app?

**Comment by wenzhiwang** [comment]
@dikaiosune It works fine until I enabled Debug Remote JS. It will disconnect right after connection. It won't try to reconnect and disconnect again like the original issue I raised; however, I cannot debug the code with Chrome debugger which is a big loss.

**Comment by wenzhiwang** [comment]
@dikaiosune  I tried to run the same code on Android simulator with Debug Remote JS enabled, socket.io connects fine. Is this fixable?

**Comment by dikaiosune** [comment]
Can you rephrase what the exact issue is? I'm not sure I follow.

**Comment by wenzhiwang** [comment]
Sorry about the confusing. I ran my code on Expo iPhone app. The original issue is that socket.io keeps disconnecting and reconnecting when I had the socket.io url pointing to my local server. I fixed the issue by changing the socket.io url to a ngrok url. It starts working fine. However, after I shake my phone and turned on Debug Remote JS, socket.io disconnect immediately after the initial connection. After I turned off Debug Remote JS and refresh, socket.io connects the server normally. I also tried other things:
a. Run the same code on Android Studio's android simulator with npm run android. Socket.io works fine regardless of turning on Debug Remote JS.
b. Run the same code on Expo Android app. Socket.io disconnects after turning on Debug Remote JS.

**Comment by ide** [comment]
Can you please make a table of the situations in which the socket.io connection works or doesn't work i.e. what mobile OS, simulator/device, Expo/not Expo? Also just to set expectations, if the websocket code works on iPhone there's guarantee it works on Android and vice versa, they are completely different implementations with different debugging internals.

**Comment by wenzhiwang** [comment]
Sure. From my observation, Socket.io won't work if the app is running on Expo with Debug Remote JS enabled. See the table. Also, if you try running this code with CRNA, you should see the issue.
<table>
<thead>
<tr>
<td>Platform</td>
<td>Expo</td>
<td>Enable Debug Remote JS&nbsp;</td>
<td>Socket.io</td>
</tr>
</thead>
<tbody>
<tr>
<td><span style="color: #ff0000;">iPhone 7+</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">&nbsp;Yes</span></td>
<td><span style="color: #ff0000;">Not working</span></td>
</tr>
<tr>
<td>iPhone 7+</td>
<td>Yes</td>
<td>No</td>
<td>Working</td>
</tr>
<tr>
<td>Android Studio Emulator</td>
<td>No</td>
<td>Yes</td>
<td>Working</td>
</tr>
<tr>
<td>Android Studio Emulator</td>
<td>No</td>
<td>&nbsp;No</td>
<td>Working</td>
</tr>
<tr>
<td><span style="color: #ff0000;">Memu's Android emulator</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">Not working</span></td>
</tr>
<tr>
<td>Memu's Android emulator</td>
<td>Yes</td>
<td>&nbsp;No</td>
<td>Working</td>
</tr>
</tbody>
</table>

**Comment by dikaiosune** [comment]
Thanks! Do you have a way to test if iPhone + not-Expo + remote debug works with socket.io? What about Android Studio emulator + Expo + remote debug (cc @jesseruder as I think there may be a little setup needed to use the AS emulator with Expo)?

**Comment by jesseruder** [comment]
There shouldn't be any problem running on the Android Studio emulator! As long as `adb` is on your path it should work.

**Comment by wenzhiwang** [comment]
@dikaiosune @ide Interesting! It's working on Mac even if Debug Remote JS is enabled. I updated the table.
<table>
<thead>
<tr>
<td>OS</td>
<td>Platform</td>
<td>Expo</td>
<td>Enable Debug Remote JS&nbsp;</td>
<td>Socket.io</td>
</tr>
</thead>
<tbody>
<tr>
<td>Windows</td>
<td><span style="color: #ff0000;">iPhone 7+</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">&nbsp;Yes</span></td>
<td><span style="color: #ff0000;">Not working</span></td>
</tr>
<tr>
<td>Windows</td>
<td>iPhone 7+</td>
<td>Yes</td>
<td>No</td>
<td>Working</td>
</tr>
<tr>
<td>Windows</td>
<td>Android Studio Emulator</td>
<td>No</td>
<td>Yes</td>
<td>Working</td>
</tr>
<tr>
<td>Windows</td>
<td>Android Studio Emulator</td>
<td>No</td>
<td>&nbsp;No</td>
<td>Working</td>
</tr>
<tr>
<td>Windows</td>
<td><span style="color: #ff0000;">Memu's Android emulator</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">Yes</span></td>
<td><span style="color: #ff0000;">Not working</span></td>
</tr>
<tr>
<td>Windows</td>
<td>Memu's Android emulator</td>
<td>Yes</td>
<td>No</td>
<td>Working</td>
</tr>
<tr>
<td>MacOS</td>
<td>iPhone 7+</td>
<td>Yes</td>
<td>Yes</td>
<td>Working</td>
</tr>
<tr>
<td>MacOS</td>
<td>iPhone 7+</td>
<td>Yes</td>
<td>No</td>
<td>Working</td>
</tr>
<tr>
<td>MacOS</td>
<td>iPhone Emulator</td>
<td>Yes</td>
<td>Yes</td>
<td>Working</td>
</tr>
<tr>
<td>MacOS</td>
<td>iPhone Emulator</td>
<td>Yes</td>
<td>No</td>
<td>Working</td>
</tr>
</tbody>
</table>

**Comment by gavin-gmlab** [comment]
I'm also facing this problem on my app. Client keeps disconnecting. Reconnection happens randomly, or when some other network (non websocket) event happens

**Comment by wenzhiwang** [comment]
@gavin-gmlab Is your OS Windows or MacOS? To me, it only happens on Windows when "debug remote JS" is enabled. Also, it seems to work with native WebSocket from FB's RN example, but it doesn't work with socket.io.  I'm also not sure if this issue only happens to the app built by CRNA.

**Comment by franckmaurin** [comment]
Hi ! I have the same issue that @gavin-gmlab, I am on MacOS and my project don't use WebSocket.

**Comment by brentvatne** [comment]
upstream react-native issue, closing
